src/Backend.idr

```diff
@@ -12,86 +12,49 @@ import Text.PrettyPrint.WL
+  typeSource : type -> Doc
 
-||| A variable environment. Left=free, Right=bound.
-Env : Nat -> Type
-Env k = Vect k (Either String Decl)
-
-||| Standard implmementation of `freshEnv` for languages that require type parameters to
-||| start with lower case letters (and allow numbers in names).
-freshEnvLC : (n : Nat) -> Env n
-freshEnvLC n = unindex {n} (\f => Left ("x" ++ show (finToInteger f)))
-
-||| Standard implmementation of `freshEnv` for languages that require type parameters to
-||| start with upper case letters (and allow numbers in names).
-freshEnvUC : (n : Nat) -> Env n
-freshEnvUC n = unindex {n} (\f => Left ("X" ++ show (finToInteger f)))
-
-||| Get the free names from the environment.
-getFreeVars : (e : Env n) -> Vect (fst (Vect.filter Either.isLeft e)) String
-getFreeVars e with (filter isLeft e)
-  | (p ** v) = map (either id (const "")) v
-
-||| Get a list of the de Brujin indices that are actually used in a `TDef`.
-getUsedIndices : TDef n -> List (Fin n)
-getUsedIndices T0         = []
-getUsedIndices T1         = []
-getUsedIndices (TSum xs)  = assert_total $ concatMap getUsedIndices xs
-getUsedIndices (TProd xs) = assert_total $ concatMap getUsedIndices xs
-getUsedIndices (TVar i)   = [i]
-getUsedIndices (TMu _ xs) = assert_total $ concatMap ((concatMap weedOutZero) . getUsedIndices . snd) xs
-  where weedOutZero : Fin (S n) -> List (Fin n)
-        weedOutZero FZ     = []
-        weedOutZero (FS i) = [i]
-getUsedIndices (TName _ t) = getUsedIndices t
-
-||| Filter out the entries in an `Env` that is referred to by a `TDef`.
-getUsedVars : Env n -> (td: TDef n) -> Env (length (getUsedIndices td))
-getUsedVars e td = map (flip index e) (fromList $ getUsedIndices td)
-
-||| Interface for codegens. `lang` is a type representing (the syntactic structure of)
-||| a type declaration in the target language.
-interface Backend lang where
-  ||| Given a `TDef` and a matching environment, generate a list of type definitions
-  ||| in the target language.
-  generateTyDefs : Env n -> TDef n -> List lang
-
-  ||| Given a type definition in the target language, generate its code.
-  generateCode : lang -> Doc
-
-  ||| Generate a new environment with n free names.
-  freshEnv : (n: Nat) -> Env n
+  ||| Generate source code for a type definition.
+  defSource  : def -> Doc
 
 ||| Use the given backend to generate code for a type definition and all its dependencies.
-generate : (lang: Type) -> Backend lang => {n: Nat} -> TDef n -> Doc
-generate lang {n} td = vsep2 . map (generateCode) . generateTyDefs {lang} (freshEnv {lang} n) $ td
-
-||| Interface for codegens which distinguish between the message type itself and
-||| its helper definitions. Currently doesn't support typedefs with free variables. 
-interface NewBackend def type | def where
-  ||| Given a `TDef`, generate its corresponding type signature.
-  msgType  : TDef 0 -> type
+generateDefs : (def: Type) -> ASTGen def type n => CodegenIndep def type => TNamed n -> Doc
```

Comment:
Constraints can be uncurried to `(ASTGen def type n, CodegenIndep def type) =>`, probably a bit more structured this way?